### PR TITLE
Add RazorLanguageVersion.3_0

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/Razor/issues/2612

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/AssemblyAttributeInjectionPass.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X
+{
+    public class AssemblyAttributeInjectionPass : IntermediateNodePassBase, IRazorOptimizationPass
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/AssemblyAttributeInjectionPass.cs]: Copied over from the `Microsoft.AspNetCore.Mvc.Razor.Extensions`.

File: build/MPack.targets [hunk 2]
```diff
@@ -73,6 +73,7 @@
 
     <ItemGroup>
       <LanguageServiceExtensionAssembly Include="$(RepositoryRoot)src\Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X\bin\$(Configuration)\net46\Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X.dll" />
+      <LanguageServiceExtensionAssembly Include="$(RepositoryRoot)src\Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X\bin\$(Configuration)\net46\Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X.dll" />
```
Comment (pranavkm) [inline build/MPack.targets]: cc @NTaylorMullen

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/Microsoft.AspNetCore.Mvc.Razor.Extensions.csproj [hunk 3]
```diff
@@ -12,11 +12,6 @@
     </Compile>
   </ItemGroup>
 
-  <ItemGroup>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/Microsoft.AspNetCore.Mvc.Razor.Extensions.csproj]: The extensions project no longer has any build content

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -5,8 +5,7 @@
 using Microsoft.AspNetCore.Mvc.Razor.Extensions;
 using Microsoft.AspNetCore.Razor.Language;
 
-[assembly: ProvideRazorExtensionInitializer("MVC-2.0", typeof(ExtensionInitializer))]
-[assembly: ProvideRazorExtensionInitializer("MVC-2.1", typeof(ExtensionInitializer))]
+[assembly: ProvideRazorExtensionInitializer("MVC-3.0", typeof(ExtensionInitializer))]
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/Properties/AssemblyInfo.cs]: Adding MVC 3-0 configuration that the compiler defaults to

File: src/Microsoft.AspNetCore.Razor.Language/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -3,25 +3,28 @@
...
+[assembly: InternalsVisibleTo("Microsoft.VisualStudio.RazorExtension, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc.Razor.Extensions.Test, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Language/Properties/AssemblyInfo.cs]: Grouped and sorted in to src, tests and perf

File: src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs [hunk 6]
```diff
@@ -30,14 +32,19 @@ public static bool TryParse(string languageVersion, out RazorLanguageVersion ver
 
             if (string.Equals(languageVersion, "latest", StringComparison.OrdinalIgnoreCase))
             {
-                version = Version_2_1;
+                version = Latest;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Language/RazorLanguageVersion.cs]: Buggy

File: src/Microsoft.VisualStudio.Editor.Razor/LegacyProjectEngineFactory_3_0.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+using System;
+using System.Reflection;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    [ExportCustomProjectEngineFactory("MVC-3.0", SupportsSerialization = true)]
+    internal class LegacyProjectEngineFactory_3_0 : IProjectEngineFactory
```
Comment (pranavkm) [inline src/Microsoft.VisualStudio.Editor.Razor/LegacyProjectEngineFactory_3_0.cs]: I'm not entirely sure why these things are named this way, but it does the job

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test.csproj [hunk 8]
```diff
@@ -41,11 +41,11 @@
     <PackageReference Include="System.Diagnostics.DiagnosticSource" Version="$(SystemDiagnosticsDiagnosticSourcePackageVersion)" />
   </ItemGroup>
 
-  <ItemGroup>
-    <MvcShim Condition="$(TargetFramework) != ''" Include="..\Microsoft.AspNetCore.Razor.Test.MvcShim\bin\$(Configuration)\$(TargetFramework)\Microsoft.AspNetCore.Razor.Test.MvcShim.deps.json" />
-  </ItemGroup>
-
   <Target Name="CopyDepsFromShims" AfterTargets="Build">
+    <ItemGroup>
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test.csproj]: Cleanup: Moved this in to the target

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X.Test/InjectDirectiveTest.cs [hunk 9]
```diff
@@ -0,0 +1,225 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text;
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X.Test/InjectDirectiveTest.cs]: Copied from `Microsoft.AspNetCore.Mvc.Razor.Extensions.Test`

File: tooling/Microsoft.VisualStudio.Mac.RazorAddin/Properties/_Manifest.addin.xml [hunk 10]
```diff
@@ -5,6 +5,7 @@
     <Import assembly="Microsoft.CodeAnalysis.Razor.Workspaces.dll" />
     <Import assembly="Microsoft.AspNetCore.Razor.Language.dll" />
     <Import assembly="Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X.dll" />
+    <Import assembly="Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X.dll" />
```
Comment (pranavkm) [inline tooling/Microsoft.VisualStudio.Mac.RazorAddin/Properties/_Manifest.addin.xml]: cc @NTaylorMullen

File: tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest [hunk 11]
```diff
@@ -26,6 +26,7 @@
     <Asset Type="Microsoft.ServiceHub.Service" d:Source="File" Path="razorLanguageService64.servicehub.service.json" />
     <Asset Type="Microsoft.VisualStudio.Assembly" Path="Microsoft.AspNetCore.Mvc.Razor.Extensions.dll" />
     <Asset Type="Microsoft.VisualStudio.Assembly" Path="Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X.dll" />
+    <Asset Type="Microsoft.VisualStudio.Assembly" Path="Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X.dll" />
```
Comment (pranavkm) [inline tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest]: How do I test this in VS - install the VSIX and try a 2.1 project?

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/MvcRazorTemplateEngine.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+using System.IO;
+using System.Text;
+using Microsoft.AspNetCore.Razor.Language;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X
+{
+    /// <summary>
+    /// A <see cref="RazorTemplateEngine"/> for Mvc Razor views.
+    /// </summary>
+    public class MvcRazorTemplateEngine : RazorTemplateEngine
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/MvcRazorTemplateEngine.cs]: hmmm. I'm not sure that we need this. Is this used anywhere?

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/RazorExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+            builder.Features.Add(new MvcViewDocumentClassifierPass());
+            builder.Features.Add(new AssemblyAttributeInjectionPass());
+            builder.Features.Add(new InstrumentationPass());
+
+            builder.SetImportFeature(new MvcImportProjectFeature());
+        }
+
+        #region Obsolete
+        [Obsolete("This method is obsolete and will be removed in a future version.")]
+        public static void Register(IRazorEngineBuilder builder)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/RazorExtensions.cs]: Can remove this I think.

File: tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest (hunk 11 above)
Comment (rynowak) [reply tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest]: you can just F5 the VSIX project and then try and 3.0 and 2.1 project

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/MvcRazorTemplateEngine.cs (hunk 12 above)
Comment (NTaylorMullen) [reply src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X/MvcRazorTemplateEngine.cs]: Pretty sure we can nuke

**Pull request merged**